// react-native-tableview-simple/lib/commonjs/components/Theme.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThemeContext = exports.THEMES = void 0;
const react_1 = __importDefault(require("react"));
exports.THEMES = {
    appearances: {
        light: {
            colors: {
                background: '#FFF',
                muted: '#C7C7CC',
                separatorColor: '#C8C7CC',
                body: '#000',
                primary: '#007AFF',
                secondary: '#8E8E93',
            },
        },
        dark: {
            colors: {
                background: '#121312',
                muted: '#59595d',
                separatorColor: '#2b2b2d',
                body: '#FFF',
                primary: '#0f64ee',
                secondary: '#aeaeae',
            },
        },
    },
};
exports.ThemeContext = react_1.default.createContext(exports.THEMES.appearances.light);
